### ISSUE TITLE
Possible visualization of expression evaluation

### ISSUE THREAD

**Issue by hkaiser** [STEllAR-GROUP/phylanx#125]
This is meant to be a sketch of an idea how we could visualize the evaluation of an execution tree. Let's assume we have a simple expression `store(A, add(B, C))`, then the following visualization allows to quickly understand what part of the expression evaluation took how long:

![expr](https://user-images.githubusercontent.com/568063/33807778-757e8816-dda1-11e7-8e9b-c64dbcd6a47a.png)

An implementation of such a visualization would require to add some functionalities:

 - [x] associate a name with every instantiated primitive (possibly derived from the source code line the primitive was created from) (see #126)
 - [x] Generate topology of execution tree in a commonly used string format (see #148, #154)
 - [x] (optionally) measure the execution counts and times for each primitive (instrument `primitive::eval` and `primitive::eval_direct` actions), this could be done using performance counters (see #96)

@kisaacs, @khuck please add more tasks to this list from your perspective

Later on, the infrastructure built for this may be used to adaptively change the evaluation of parts of the execution tree by dynamically switching to direct execution in cases where the asynchronous (default) execution may impose too much overheads (see #121).

**Comment by kisaacs** [comment]
It sounds like you know what data we need (tree specification, correspondence to code/variable names, metrics). Who do we see as the user for these? Phylanx developers or phylanx users? 

If it's developers first, as a first cut, we could write a script that outputs a static PDF -- this might match the quick command-line workflow (or even we could probably output them to the command line as ASCII or interactive ASCII with ncurses). If we find we want to look at more than one metric and we want something more general user-friendly, then we might want something like the Perfopticon visualization for distributed databases:

https://idl.cs.washington.edu/papers/perfopticon/
Perfopticon keeps the tree/graph simple and augments with quick interactive chart support.

We could also alter the tree drawing itself such that the length of the edges/nodes encodes time or counts. Some of these choices will be affected by how big we think these expressions are going to get.

**Comment by hkaiser** [comment]
> It sounds like you know what data we need (tree specification, correspondence to code/variable names, metrics). Who do we see as the user for these? Phylanx developers or phylanx users?

For now it's mostly us developers being interested in getting our performance straight. Later on users might be interested as well, I'm not sure if that's too low level for them.

> Perfopticon keeps the tree/graph simple and augments with quick interactive chart support.

Looks nice.

> Some of these choices will be affected by how big we think these expressions are going to get.

The trees might become very large, so some means of aggregation of sub-trees during visualization might be needed as well.

**Comment by hkaiser** [comment]
Another thoughts: 

- Our trees are to be evaluated concurrently (i.e. sub-trees of the same node might 'run' in parallel). It might be interesting to somehow visualize the amount of parallelism achieved while evaluating the tree. -- Also visualizing the critical path of evaluating the tree (i.e. what path from the root to any of the leaves takes the longest time to evaluate) might be useful.

**Comment by stevenrbrandt** [comment]
On Thu, Dec 14, 2017 at 9:46 PM, Kate Isaacs <notifications@github.com>
wrote:

> For aggregation and critical path overlays, interactivity would be helpful
> so web-based is probably the way to go over command line.
>
Jupyter notebooks let you have web-based interactivity *and* a command line.

It also has built-in visualzation capabilities based on matplotlib.

--Steve

**Comment by khuck** [comment]
Interestingly, Nick Chaimov is rewriting some TAU viz tools with jupyterlab, and getting some very promising results.  But not all charting libraries in Python are created equal, and getting an interactive output is not that straightforward.  Matplotlib, for example, only outputs static images.

I don't know what's available for interactive viz of graphs/trees in Python, though.

**Comment by stevenrbrandt** [comment]
Actually, matplotlib also does movies. In addition, it's possible to get
Jupyter + matplotlib to regenerate images on the fly in response to
sliders, etc.

--Steve

**Comment by ct-clmsn** [comment]
@kisaacs perfopticon looks like an excellent starting point. A couple of web-based (python/javascript) implementation technologies:

[D3](https://d3js.org/)
[Bokeh](http://bokeh.pydata.org/en/latest/)
[plotly](https://plot.ly/) - plotly provides some python and javascript development tools ie: [Dash](https://plot.ly/products/dash/), [plotly.js](https://plot.ly/plotly-js-scientific-d3-charting-library/), [plotly.py](https://plot.ly/d3-js-for-python-and-pandas-charts/), [plotly.ggplot](https://plot.ly/ggplot2/)

As far as specific types of visualizations, [tensorboard](https://github.com/tensorflow/tensorboard) has some interesting ideas for [graph vis](https://www.tensorflow.org/get_started/graph_viz) and [temporal/histogram rendering](https://www.tensorflow.org/get_started/tensorboard_histograms).

Given we're looking at a server based system, would ya'll expect compute graphs and trace data would get pushed to the server from both the python interface (so user from a notebook) when a task is posted, and from the execution engine when the computation completes?

**Comment by ct-clmsn** [comment]
@hkaiser 

> Our trees are to be evaluated concurrently (i.e. sub-trees of the same node might 'run' in parallel). It might be interesting to somehow visualize the amount of parallelism achieved while evaluating the tree.

That would be a nice experiment - every critical path node pushes a "progress" message to the visualization server.

**Comment by khuck** [comment]
Bokeh is what Nick is using (so far).

@ct-clmsn  The TAU tools are 10-15 year old Java implementations.  It's long, long overdue.  We've talked about it for years, but it's not "research", so...

**Comment by ct-clmsn** [comment]
@khuck nice, that helps sort out the space. measuring the community support for visualization tools is no easy task. would you want to push information about APEX's policy choices out to the visualization tool (Subtree/node X met the criteria for an APEX policy to engage)?

**Comment by kisaacs** [comment]
@stevenrbrandt I was thinking of developers using the shell--sorry, should have been more clear. I don't think they're doing their shell access through Jupyter, though they could. How Jupyter-workflows may change the visualization-interaction process is something I've been thinking about for a while, but I don't think we have the users right now. I'm more concerned with direct manipulation of node-link diagrams and timelines in terms of interactivity. I've done interactive matplotlib visualizations before and I think for this problem there are better alternatives.

@ct-clmsn I foresee us using d3js as you listed. Since we'll likely need to get into the nitty-gritty to customize the design, it gives us a lot of low-level flexibility and is extremely well-documented with lots of examples. I think at first, we should aim for post-mortem analysis, because we need a really good reason to add overhead and I'm guessing the analysis timeline for developers is not on the scale of 'while it runs'. If we do get to the point of steering, that's probably going to be a different visualization design which then gracefully falls into whatever we design for post-mortem analysis.

It sounds like the limiting factor in most of the proposals is still having the tree data--so again, any way to get something close to that, even if it doesn't have all the features we want right now?

**Comment by hkaiser** [comment]
@kisaacs in what format would you like to have the tree (connectivity) data in? Are there any 'standard' formats normally used for this?

**Comment by kisaacs** [comment]
@hkaiser The dot format (what GraphViz uses) is common for visualizing graphs which has a side benefit of always being able to just run dot. More compact formats exist for big data graph processing, especially since trees are sparse--we can always convert. If we're sure it will be a tree, the Newick format is pretty compact.

**Comment by khuck** [comment]
@hkaiser @kisaacs APEX already has code in place to generate this, we just have to make the calls.  All that is needed is to add this API call during a task:
```
/**
 \brief Create a new task (dependency).

 This function will note a task dependency between the current 
 timer (task) and the new task.

 \param name The name of the timer.                                        
 \param task_id The ID of the task 
 \return No return value.
 */
APEX_EXPORT void new_task(const std::string &name, uint64_t task_id);
```

Then run with the APEX_TASKGRAPH_OUTPUT environment variable set to 1.

The question is "where to put the apex::new_task() call?"

**Comment by hkaiser** [comment]
@khuck should we really conflate the task API exposed by APEX (that is meant for tracking HPX tasks, right?) with the logical execution tree structure (that might contain loops etc., i.e. single nodes/sub-trees that would create a lot of HPX tasks while being executed)?

**Comment by khuck** [comment]
@hkaiser probably not.  But @kisaacs was asking for data... :)

**Comment by kisaacs** [comment]
Manually instrumenting an example might give us a better idea of what we really want/need before making the effort to code it in. I guess that means I should prioritize building Phylanx after all. Is there a suggested started application and part of the API we want to target for adding these calls?

**Comment by hkaiser** [comment]
There is the LRA example (https://github.com/STEllAR-GROUP/phylanx/tree/master/examples/algorithms). I also think that @parsa should close in to having the first performance counters in place for us to be able to collect some date. I will also work on providing you with a means of traversing the execution tree soon.

**Comment by kisaacs** [comment]
@hkaiser I am not familiar with how the Newick format handles attributes. (@kawilliams might... can I get her added to this group/repository?) With the dot format, you can add attributes similar to HTML tag attributes. Libraries exist for accessing them. However, while it means keeping track of separate files, I like the table of attribute idea (like the CSV you suggest) assuming we have node IDs to do that matching. That would probably help with related views (e.g., statistical plots) where a tabular base makes more sense. 

Ultimately if there's a performance concern in writing out the perf data, we can always do the data wrangling after the fact assuming post-mortem analysis.

**Comment by khuck** [comment]
@hkaiser, @kisaacs The "dot" format is pretty widely used (from what I understand) and it allows multiple attributes for the graph, edges and nodes. However, as we are building the visualizer, I don't think we have to constrain ourselves to a specific format. "dot" is used for general graphs, but is a tree representation sufficient?

**Comment by hkaiser** [comment]
@kisaacs I have sent an invitation to join this group to @kawilliams.

> Ultimately if there's a performance concern in writing out the perf data, we can always do the data wrangling after the fact assuming post-mortem analysis.

Even then we'd need to somehow write the information to disk...

**Comment by hkaiser** [comment]
@khuck, @kisaacs: also, I will add generating the tree topology in dot format, that should give us more options.

**Comment by hkaiser** [comment]
@khuck from what I read (here: http://www.graphviz.org/doc/info/attrs.html), we can't simply add arbitrary attributes to the node description in the dot format, The attributes are restricted to the described set targeting mainly visualization properties. Is that correct?

**Comment by hkaiser** [comment]
> @khuck, @kisaacs: also, I will add generating the tree topology in dot format, that should give us more options (see #153, Generate expression tree topology in DOT format).

Done.

**Comment by hkaiser** [comment]
I think this can be closed now.
